parse botocore.endpoint logs into a list of AWS actions

s-hertel opened the pull request:
Add an option to parse botocore.endpoint logs for the AWS actions performed during a task

Add a callback to consolidate all AWS actions used by modules

Allow integration tests to output AWS actions

Added some docs to help contributors help with the needs_ci_update queue

##### SUMMARY
This helps remove some friction when determining the minimum permissions needed for integration tests, as well as being a step in the direction of being able to document https://github.com/ansible/ansible/issues/22552

Please bikeshed names.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

s-hertel committed:
Add an option to parse botocore.endpoint logs for the AWS actions performed during a task

Add a callback to consolidate all AWS actions used by modules

Allow integration tests to output AWS actions

Added some docs

s-hertel commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -120,14 +128,38 @@ def __init__(self, **kwargs):
+            if found_operational_request:
+                operation_request = found_operational_request.group(0)[20:-1]
+                resource = re.search("https://.*?\.", ln).group(0)[8:-1]
+                actions.append("{0}:{1}".format(resource, operation_request))
```
It would be nice if this was a list of (action, resource_arn) pairs, but I'm still trying to find any ARN logging

s-hertel replied:
@mattclay Searching to no avail, the closest I can get to generating resource restrictions is by scraping the AWS documentation. The targeted resources are not logged and I can't find a way to construct them from anything in botocore.

Here's a script that takes the output from this callback and creates a policy with the documented resource restrictions: https://gist.github.com/s-hertel/00ffb9285c417c5955eab63a042a8c32.

s-hertel committed:
Tidy docs and code

s-hertel committed:
Make code a little more flexible

mattclay commented on test/runner/lib/executor.py:
```diff
@@ -1115,7 +1115,7 @@ def integration_environment(args, target, cmd, test_dir):
     integration = dict(
         JUNIT_OUTPUT_DIR=os.path.abspath('test/results/junit'),
-        ANSIBLE_CALLBACK_WHITELIST='junit',
+        ANSIBLE_CALLBACK_WHITELIST='junit, aws_resource_actions',
```
This should only be enabled for AWS tests. Append it to the environment variable here instead:

https://github.com/ansible/ansible/blob/332dc6a93a1d791270fba012c8ebb28f3a70a81f/test/runner/lib/cloud/aws.py#L87

mattclay commented on test/integration/targets/route53/tasks/main.yml:
```diff
@@ -9,6 +9,7 @@
 - name: Test basics (new zone, A and AAAA records)
   module_defaults:
     group/aws:
+      debug_botocore_endpoint_logs: True
```
Can this be safely enabled for all AWS tasks in our tests?

If so, ansible-test could be extended so cloud plugins could set module defaults at the play level instead.

mattclay replied:
To extend ansible-test to enable setting defaults at the task level, you'll need to add another method to the base class that can return module defaults:

https://github.com/ansible/ansible/blob/f97abc095b244038b4e4d59fb7919ba15643c230/test/runner/lib/cloud/__init__.py#L378

Then add that method to the aws.py cloud plugin to return the needed setting.

Call the method and use it when generating the playbook:

https://github.com/ansible/ansible/blob/f97abc095b244038b4e4d59fb7919ba15643c230/test/runner/lib/executor.py#L1185-L1195

Note that this won't work for script based tests since they don't use our generated playbook.

mattclay commented on lib/ansible/plugins/callback/aws_resource_actions.py:
```diff
@@ -0,0 +1,72 @@
+
+    def runner_on_ok(self, host, res):
+        if res.get('resource_actions'):
+            self.aws_resource_actions.extend(res['resource_actions'])
```
Refactor these two lines into a function and call it from each of the 4 event handler functions.

mattclay commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -120,14 +128,38 @@ def __init__(self, **kwargs):
+
     def exit_json(self, *args, **kwargs):
+        if self.params.get('debug_botocore_endpoint_logs'):
+            kwargs['resource_actions'] = self._get_resource_action_list()
```
Refactor these two lines into a function and call it from the exit_json, fail_json and fail_json_aws.

Maybe just move the logic into _get_resource_action_list and rename it.

s-hertel replied:
I refactored but didn't move the logic into `_get_resource_action_list`because it was intended to collect unique actions an individual task uses. Not strictly necessary for collecting actions a given playbook uses, but I thought having the finer grain of control had merit.

s-hertel committed:
Enable aws_resource_actions callback only for AWS tests

Allow module_defaults to be set at the play level for cloud targets

s-hertel committed:
Refactor callback

s-hertel commented:
@mattclay I've made all of your requested changes, except part of https://github.com/ansible/ansible/pull/49312#discussion_r249583883. I'm setting module_defaults at the play level for AWS modules using the test executor, but there is a bug on devel that overwrites module_defaults if they are also set at a task level (which many of the AWS tests do) rather than merging the variables. I have a bugfix open for that here: https://github.com/ansible/ansible/pull/51149.

When I fix/document the script you took a look at before, do you think it's alright if I place that in the hacking/ directory?

mattclay commented:
@s-hertel Yes, I'd put that script in the hacking directory.

mattclay commented on test/runner/lib/executor.py:
```diff
@@ -1195,13 +1195,15 @@ def command_integration_role(args, target, start_at_task, test_dir, inventory_pa
 
         if cloud_environment:
             hosts = cloud_environment.inventory_hosts or hosts
+            module_defaults = cloud_environment.get_play_level_module_defaults()
```
A default vaule for `module_defaults ` is needed when not testing a cloud environment.

Lacking this is the cause of the current integration test failures.

s-hertel committed:
pylint fixes

s-hertel committed:
Add script to help generate policies

s-hertel committed:
pep8

s-hertel committed:
remove duplicate key and put keys in alphabetical order to help avoid that in the future

s-hertel committed:
fix script for duplicates

s-hertel commented:
Rebased due to module_doc_fragments changing locations. @mattclay I believe I've addressed all of your comments.

mattclay commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -263,6 +295,14 @@ def __getattr__(self, name):
             return unwrapped
 
 
+def inject_botocore_logged_actions(method):
```
What is this method used for? I don't see any references to it.

s-hertel replied:
Oh, good catch. I was going to decorate exit_json, fail_json and fail_json_aws, decided against it midway through and then forgot to remove it.

mattclay commented on test/runner/lib/cloud/__init__.py:
```diff
@@ -410,3 +410,9 @@ def inventory_hosts(self):
+
+    def get_play_level_module_defaults(self):
+        """Return a dictionary of group(s) containing vars to set for the whole test target.
+        :rtype: dict | None
```
```suggestion
        :rtype: dict[str, dict[str, any]] | None
```

mattclay commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
```diff
@@ -653,3 +653,68 @@ integration builds, so the tests will initially fail when you submit a pull requ
+
+The simplest way is to run the tests using ample permissions with the `debug_botocore_endpoint_logs` module option for AnsibleAWSModule modules.
+After the `PLAY RECAP` of the completed tests the list of AWS actions that need to be allowed will be displayed.
+1) Ensure you set the `debug_botocore_endpoint_logs` module option set to True in your tests.
```
This step is no longer needed now that it's integrated into the cloud module for ansible-test.

s-hertel committed:
Remove unused decorator

s-hertel committed:
Remove step in guidelines

s-hertel commented:
Rebased for conflict in module_utils/aws/core.py

mattclay commented:
The changes here look good. Once https://github.com/ansible/ansible/pull/53044/ is merged this PR can be rebased to take advantage of the new `module_defaults` and `callback_plugins` arguments available for cloud plugins. No changes to `test/runner/executor.py` should be needed then.

mattclay commented:
@s-hertel The cloud plugin overhaul has been merged.

s-hertel committed:
Fix setting module_defaults and the callback plugin for the test runner

s-hertel commented on test/integration/targets/ec2_group/tasks/main.yml:
```diff
@@ -11,6 +11,7 @@
 - include: ./credential_tests.yml
 - module_defaults:
     group/aws:
+      debug_botocore_endpoint_logs: True
```
Once #51149 is merged I'll be able to remove this. Right now any tests setting module_defaults for group/aws (such as this one) overwrite the setting from the test runner.

s-hertel committed:
Set debug_botocore_endpoint_logs via environment variable instead

Ensure AWS tests inherit environment

Remove AWS CLI in aws_rds inventory tests and use the module

s-hertel commented:
bot_status

acozine committed:
first edits

minimize and clarify

revise testing procedures

moves troubleshooting into its own section

this is cleaner

bullets for the two choices

make requirements clearer

s-hertel commented:
!needs_revision
